docs/content/guides/dagster/building-and-memoizing-versioned-assets.mdx

```diff
@@ -0,0 +1,191 @@
+
+Note that in the created run, _only_ the step associated with `multipled_number` is run. The system knows that `versioned_number` is up to date and therefore can safely skip that computation.
+
+Now let's update the `versioned_number` asset-- change its return value and code version. As we saw previously, this will cause `versioned_number` to become stale. But since `multipled_number` depends on `versioned_number`, it must be recomputed as well, and so is also marked stale:
+
+<Image
+alt="Dependencies code version only"
+src="/images/guides/building-versioned-assets/dependencies-code-version-only.png"
+width={2662}
+height={1748}
+/>
+
+
+Click "Materialize stale and missing" to get both assets up-to-date again.
+
+---
+
+## Step Three: Staleness with Source Assets
+
+In the real world, data pipelines depend on external upstream data. So far in this tutorial, we haven't used any external data-- we've been substituting hardcoded data in the asset at the root of our graph, and using a code version as a stand-in for the version of that data. We can do better than this.
+
+External data sources in Dagster are modeled by <PyObject object="SourceAsset" displayText="SourceAssets" />. We can add versioning to a `SourceAsset` by making it _observable_. An observable source asset has a user-defined function that computes a logical version.
+
+Let's add an <PyObject object="observable_source_asset" decorator="true" /> called `input_number`. This will represent a file written by an external process, upstream of our pipeline. The body of the `input_number` function computes a hash of the file contents and returns it as a `LogicalVersion`. We'll set `input_number` as an upstream dependency of `versioned_number`.
+
+```python file=/guides/dagster/building_versioned_assets/observable_source_asset_path_with_non_argument_deps.py
+from hashlib import sha256
+
+from dagster import LogicalVersion, asset, file_relative_path, observable_source_asset
+
+def sha256_digest_from_str(string: str) -> str:
+    hash_sig = sha256()
+    hash_sig.update(bytearray(string, "utf8"))
+    return hash_sig.hexdigest()
+
+
+FILE_PATH = file_relative_path(__file__, "input_number.txt")
+
+
+@observable_source_asset
+def input_number():
+    with open(FILE_PATH) as ff:
+        return LogicalVersion(sha256_digest_from_str(ff.read()))
+
+
+@asset(code_version="v3", non_argument_deps={"input_number"})
+def versioned_number():
+    with open(FILE_PATH) as ff:
+        return int(ff.read())
+
+
+@asset(code_version="v1")
+def multipled_number(versioned_number):
+    return versioned_number * 2
+```
+
+Adding an observable source asset to our graph has revealed a new button labeled "Observe Sources". We click this button to kick off a run that executes the observation function of `input_number`. Let's look at the entry in the asset catalog for `input_number`:
+
```

Comment:
Show another screenshot:

![Screen Shot 2023-02-09 at 9 48 00 AM](https://user-images.githubusercontent.com/28738937/217845913-1f2ebc8b-f7fa-404a-87b1-021de6537681.png)

Reply:
This screenshot is identical to the post-reload one above-- how about just one screenshot here, under both the code snippet and the paragraph about the `Observe sources` button